### ISSUE TITLE
NLog not flushing before application exit using NLog.Extensions.Logging

### ISSUE THREAD

**Issue by kcritchnau** [NLog/NLog#5097]
I've noticed that when I'm using NLog through Microsoft.Extensions.Logging, the end of the application doesn't log as expected when using an async wrapper with the latest version of NLog.Extensions.Logging. If I add intermediate code, it logs, up until the very end of the application (almost like something isn't "turning out the lights," so to speak). Downgrading to NLog.Extensions.Logging 1.7.5 fixes the issue, as does a Thread.Sleep or LogManager.Shutdown before the end of the application.

**NLog version**: 5.0.5
**NLog.Extensions.Logging version**: 5.1.0

**Platform**:  .NET6

**Current NLog config** (xml or C#, if relevant)

```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      autoReload="true"
      internalLogLevel="Info"
      internalLogFile="logs\internal\nlog.log">
  <variable name="internalLogDir" value="logs\internal"/>
  <variable name="baseLogDir" value="logs"/>

	<targets>
		<target
		  xsi:type="File"
		  name="BaseLog"
		  createDirs="true"
		  autoFlush="true"
		  fileName="${baseLogDir}\NLogTest.log"
		  layout="${longdate} ${callsite} ${level} ${message} ${exception:format=tostring}"
		  archiveEvery="Day"
		  archiveNumbering="Date"
		  archiveDateFormat="yyyy-MM-dd"
		  maxArchiveFiles="7"
		  concurrentWrites="true" />
		<wrapper-target xsi:type="AsyncWrapper" name="asyncBase">
			<target-ref name="BaseLog"/>
		</wrapper-target>
	</targets>
  <rules>    
    <logger name="*" minlevel="Debug" writeTo="asyncBase"/>
  </rules>
</nlog>

```
Code:
```cs
namespace NLogTest
{
    class Program
    {
        static void Main(string[] args)
        {
            var services = new ServiceCollection();
            services.AddLogging(loggingBuilder =>
            {
                // configure Logging with NLog
                loggingBuilder.ClearProviders();
                loggingBuilder.SetMinimumLevel(LogLevel.Trace);
                loggingBuilder.AddNLog("nlog.config");
            });

            var environmentName = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") ?? "Development";
            var builder = new ConfigurationBuilder()
                .SetBasePath(AppDomain.CurrentDomain.BaseDirectory)
            IConfiguration configuration = builder.Build();
            var serviceProvider = services.BuildServiceProvider();
            var logger = serviceProvider.GetService<ILogger<Program>>();
            try
            {
                logger.LogInformation("Program Start");
                Console.WriteLine("Program Start");
                logger.LogInformation("Program End");
                Console.WriteLine("Program End");
                //uncommenting either of these lines makes the logging work
                //LogManager.Shutdown();
                //Thread.Sleep(3000);
            }
            catch
            {
                Console.WriteLine("Error");
            }
        }
    }
}
```
- When running this application, the console is written to, but the log file is not. 
- Are there any workarounds? Yes, using LogManager.Shutdown() before the end of the application.

**Comment by snakefoot** [comment]
NLog hooks into AppDomain.Unload and performed flush/shutdown on these "final" events. 

Sadly enough Microsoft .NET signals these events on Linux before application is fully shutdown, thus causing the NLog logging to become silent and fails logging all events during application-shutdown.

When using NLog.Extensions.Logging v5 then it is expected that the developer cares about object-lifetime and disposes its objects for correct shutdown.

**Solution 1:**
Change this:
```c#
var serviceProvider = services.BuildServiceProvider();
```
To this (Adding `using`):
```
using var serviceProvider = services.BuildServiceProvider();
```

**Solution 2**
Introduce finally-block:
```c#
try
{
   // App Code
}
finally
{
   NLog.LogManager.Shutdown();
}
```

**Comment by kcritchnau** [comment]
This is on Windows, so the Linux behavior is not so much of a concern. Thanks for your feedback; I have implemented Solution #2 above as a work around, so if that is a recommended fix then I will just stick with that. 

Thanks again!

**Comment by snakefoot** [comment]
Yes the easy solution would be if Microsoft Dotnet worked as expected on Linux. Then NLog AutoShutdown could be enabled as before.

Sent from my Galaxy



-------- Original message --------
From: kcritchnau ***@***.***>
Date: 03/11/2022 19:11 (GMT+01:00)
To: NLog/NLog ***@***.***>
Cc: Rolf Kristensen ***@***.***>, Comment ***@***.***>
Subject: Re: [NLog/NLog] NLog not flushing before application exit using NLog.Extensions.Logging (Issue #5097)


This is on Windows, so the Linux behavior is not so much of a concern. Thanks for your feedback; I have implemented Solution #2<https://github.com/NLog/NLog/issues/2> above as a work around, so if that is a recommended fix then I will just stick with that.

Thanks again!
